pkg/virt-api/rest/subresource.go

```diff
@@ -517,6 +518,99 @@ func (app *SubresourceAPIApp) RestartVMRequestHandler(request *restful.Request,
 	response.WriteHeader(http.StatusAccepted)
 }
 
+func (app *SubresourceAPIApp) RenameVMRequestHandler(request *restful.Request, response *restful.Response) {
+	name := request.PathParameter("name")
+	namespace := request.PathParameter("namespace")
+
+	// Read the new name of the VM
+	newNameBuf, err := ioutil.ReadAll(request.Request.Body)
+
+	if err != nil {
+		writeError(errors.NewConflict(v1.Resource("virtualmachine"),
+			name, err), response)
+		return
+	}
+
+	newName := string(newNameBuf)
+
+	request.Request.Body.Close()
+
+	if name == newName {
+		writeError(errors.NewConflict(v1.Resource("virtualmachine"),
+			name, fmt.Errorf("The new name cannot be the same as the old name")), response)
+		return
+	}
+
+	vm, statusErr := app.fetchVirtualMachine(name, namespace)
+
+	/*
+		Items to copy to the new VM resource:
+		- Finalizers - Should I let finalizers run on deletion or remove them and apply them to the new VM?
+		- Initializers - Do not exist anymore since the VM is already initialized, are they needed?
+		+ OwnerReferences
+		+ Annotations
+		+ Labels
+		+ Spec
+	*/
+
+	if statusErr != nil {
+		writeError(statusErr, response)
+		return
+	}
+
+	if vm.Spec.Running == nil {
+		writeError(errors.NewInternalError(fmt.Errorf("Could not determine the running status of the vm")),
+			response)
+		return
+	}
+
+	if *vm.Spec.Running {
+		writeError(errors.NewConflict(v1.Resource("virtualmachine"),
+			name, fmt.Errorf("Renaming a running vm is not allowed")), response)
+		return
+	}
+
+	newVM := &v1.VirtualMachine{
+		TypeMeta: vm.TypeMeta,
+		ObjectMeta: k8smetav1.ObjectMeta{
+			Name:            newName,
+			Namespace:       namespace,
+			Labels:          vm.ObjectMeta.Labels,
+			Annotations:     vm.ObjectMeta.Annotations,
+			OwnerReferences: vm.ObjectMeta.OwnerReferences,
+			Finalizers:      vm.ObjectMeta.Finalizers, // Need to update the old VM and remove its finalizers and apply them here
+			//Initializers:    vm.ObjectMeta.Initializers, // Need to discover which initializers the VM had and apply them here
+		},
+		Spec: vm.Spec,
+	}
+
+	// Creating a new VM with the new name
+	newVM, err = app.virtCli.VirtualMachine(namespace).Create(newVM)
```

Comment:
The problem I see with this approach is failure handling.  What if the apiserver crashes after creating the new vm but before deleting the old one?  Maybe just have the apiserver set a field/annotation in the VM that triggers a controller to do the rest in a resilient way?

Reply:
That's a great point, I totally agree and will adapt your approach, thanks!

Reply:
@mhenriks Hi Michael, I implemented your approach, you're welcome to take a look.
If you want to wait for tests to be implemented that's fine too.
Thanks!